### PR TITLE
initial add of proposed non-func-req'ments guide

### PR DESCRIPTION
**Pull request by sbohlen**
# Non-Functional Requirements process guide

## What are you trying to address

Closing #684 

**Note**
This PR and related issue have been tagged w/ the `design-patterns` label b/c there seemed no better match.  Recommend at least considering the expansion of the label taxonomy to include e.g., "engineering process" or other so that such elements have a more 'natural' label applied.  `design-patterns` is a somewhat awkward fit for this content and is unlikely to lead to its (easy) discovery within the greater playbook.

## Description of new changes

Introduction of new reference document.

Notes/Considerations
* PR attempts to comply with the proposed folder structure outlined in PR #678 
* should this structure change (or this interpretation of the intent of that structure be incorrect) the folder structure proposed by this PR would be sim. invalid/incorrect and require correlated adjustment
* once PR #678 is completed/merged, a reference/link from the `/docs/design/design-patterns/readme.md` contained therein to the single file contained in this PR should be added
* because of this, recommendation is that this PR **not** be accepted/merged until _after_ PR #678 is merged and an edit to the resulting `readme.md` can be made as an update to this PR (else another issue should be filed to represent that proposed edit so that its need not be overlooked/lost) 

## For all pull requests

- [X] Link to the issue you are solving (so it gets closed)
- [X] Label the pull request with the appropriate area(s)
- [X] Assign potential reviewers (you may also want to contact them on Teams to ensure timely reviews)
- [X] Assign the project - Engineering Playbook Backlog
- [X] Assign the pull request to yourself

### CODE REVIEW

File: docs/design/design-patterns/non-functional-requirements_investigation_guide.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
+# Non-Functional Requirements Investigation
+
+## Background
+
+In software engineering, detailed investigation of non-functional requirements is often deferred until later stages of the project in favor of a more exclusive initial focus on the functional requirements.
+
+However, introducing non-functional requirements in the later stages of the SDLC is often highly-disruptive and inefficient to the ongoing engineering process.  Because of this, important characteristics of the system designed to provide for necessary e.g., testability, reliability, scalability, observability, securability, manageability are better considered as first-class citizens in the requirements gathering process to be defined in detail early in the engagement.
```
Comment (zmmille2) [inline docs/design/design-patterns/non-functional-requirements_investigation_guide.md:7]: nit: I prefer sentences on newlines in Markdown.

Comment (zmmille2) [reply docs/design/design-patterns/non-functional-requirements_investigation_guide.md:7]: Please expand the acronym "SDLC" before using it for the first time. I assume you mean "Software Development Life Cycle"

File: docs/design/design-patterns/non-functional-requirements_investigation_guide.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
+# Non-Functional Requirements Investigation
```
Comment (zmmille2) [inline docs/design/design-patterns/non-functional-requirements_investigation_guide.md:1]: The filename mixes use of dashes and underscores - other files under `design` prefer dashes, please update to match.

File: docs/design/design-patterns/non-functional-requirements_investigation_guide.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  * child1
+  * child2
+
+In the preceding example, 'Susan' is responsible for `Parent1` and all of its descendants _except_ for `Parent1/child2` and `Parent1/child2/grandchild1`  (for which 'John' is the stakeholder).  'Sam' is responsible for the entirety of `Parent2` and all of its descendants.
+
+This approach permits the retention of the logical hierarchy of elements themselves while also flexibly interleaving the 'stakeholder' identifications within the hierarchy of topics if/when they may need to diverge due to e.g., CUSTOMER organizational nuances.
+
+***
+
+## Areas of Investigation
```
Comment (zmmille2) [inline docs/design/design-patterns/non-functional-requirements_investigation_guide.md:47]: You mention that this section is an awkward fit under `design-patterns`, and discoverability might not be exactly what we're looking for right away. We might do well to add links to and from other parts of the playbook to drive consumption of these non-functional requirements.

In particular, I think linking to [security](https://github.com/microsoft/code-with-engineering-playbook/tree/main/docs/security) might be a good fit, as well as [observability](https://github.com/microsoft/code-with-engineering-playbook/tree/main/docs/observability).
Adding an optional section to the [Engineering Fundamentals Checklist](https://github.com/microsoft/code-with-engineering-playbook/blob/main/docs/ENG-FUNDAMENTALS-CHECKLIST.md) could also get more eyes on these requirements.

**Review by zmmille2 — CHANGES_REQUESTED** [review]
This is a great list to take into account when accommodating non-functional requirements. I think you're right that discoverability might not be where we want it, but I think there's some changes we can and should make here to try to address that.

File: docs/design/design-patterns/non-functional-requirements_investigation_guide.md (hunk 1 above)
Comment (sbohlen) [reply docs/design/design-patterns/non-functional-requirements_investigation_guide.md:7]: Can you elaborate on the comment re: newlines for sentences?  Do you literally mean _each_ sentence should be on a separate line?  Since its a non-trivial amount of editing, just want to ensure correct understanding before I proceed to make the changes.  So this preceding para in this reply would instead be....

```
Can you elaborate on the comment re: newlines for sentences?
Do you literally mean _each_ sentence should be on a separate line?
Since its a non-trivial amount of editing, just want to ensure correct understanding before I proceed to make the changes.  So this preceding para in this reply would instead be....
```

File: docs/design/design-patterns/non-functional-requirements_investigation_guide.md (hunk 3 above)
Comment (sbohlen) [reply docs/design/design-patterns/non-functional-requirements_investigation_guide.md:47]: Links added.  Also link to doc from 'root' `readme.md` in `/docs/design/` folder.  This might be preferable instead from the `readme.md` in the `/docs/design/design-patterns/` folder.   If so, please comment accordingly and will adjust.  Else, if acceptable as-is am leaving this comment OPEN for you close accordingly.

File: docs/design/design-patterns/non-functional-requirements_investigation_guide.md (hunk 1 above)
Comment (sbohlen) [reply docs/design/design-patterns/non-functional-requirements_investigation_guide.md:7]: (also, SDLC initialism explained in first-use)

File: docs/design/README.md [hunk 4]
```diff
@@ -15,3 +15,4 @@ This covers not only technical design of software, but also architecture design
 
 * [Design Patterns](design-patterns/README.md)
 * [Design Reviews](design-reviews/README.md)
+* [Non-Functional Requiurements Guidance](design-patterns/non-functional-requirements-investigation-guide.md)
```
Comment (fnocera) [inline docs/design/README.md:18]: Nit: typo "Requirements"

Comment (fnocera) [reply docs/design/README.md:18]: I also suggest linking this section from the [Sprint walkthrough](https://github.com/microsoft/code-with-engineering-playbook/blob/main/docs/SPRINT-STRUCTURE.md) as something to do before Day 1. 

We could also add to the [Engineering checklist](https://github.com/microsoft/code-with-engineering-playbook/blob/main/docs/ENG-FUNDAMENTALS-CHECKLIST.md#design-reviews) perhaps even under Design Reviews as "We have gathered clear non-functional requirements before starting the design process" 

Additionally it could definitely be added as something that should be done in the [ADS section of this document ](https://github.com/microsoft/code-with-engineering-playbook/blob/main/docs/design/design-reviews/recipes/engagement-process.md#envisioning--architecture-design-session-ads)

File: docs/design/design-patterns/non-functional-requirements-investigation-guide.md [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+While this is always a good _general_ rule to follow in software engineering, non-functional requirements for projects undertaken in highly-regulated environments (Financial Services, Gov, Healthcare) frequently have __outsized influence__ on the design and development of software systems. This makes their introduction later in the SDLC significantly more disruptive and expensive.  This increases even _further_ the importance of capturing these non-functional requirements as early in the engagement as possible when executing within a highly-regulated context.
+
+To support the process of capturing a project's _comprehensive_ non-functional requirements, this document offers a taxonomy for non-functional requirements and provides a framework for their identification, exploration, and eventual codification into formal engineering requirements as input to subsequent solution design.
+
+## Investigation Process
+
+1. Identify/brainstorm likely areas/topics requiring further investigation/definition
+1. Identify CUSTOMER stakeholder(s) responsible for each identified area/topic
```
Comment (fnocera) [inline docs/design/design-patterns/non-functional-requirements-investigation-guide.md:16]: Nit: I guess it's for emphasis but customer in caps looks a little funny, would suggest making it lower case

File: docs/design/design-patterns/non-functional-requirements-investigation-guide.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
+# Non-Functional Requirements Investigation
+
+## Background
+
+In software engineering, detailed investigation of non-functional requirements is often deferred until later stages of the project in favor of a more exclusive initial focus on the functional requirements.
```
Comment (fnocera) [inline docs/design/design-patterns/non-functional-requirements-investigation-guide.md:5]: I would suggest rewording this section to not say what we _shouldn't do_ but actually what we _should do_, so cut out the cons and lead with the positives/benefits of adopting the practice. 

I also recommend making this section more clear and concise.

File: docs/design/design-patterns/non-functional-requirements-investigation-guide.md [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+  * Services
+  * Authorities/directories
+  * Mechanisms/handshakes
+    * Active Directory
+    * SAML
+    * OAuth
+    * Other?
+  * RBAC
+    * Perms inheritance model
+* Enterprise Monitoring/Operations
```
Comment (fnocera) [inline docs/design/design-patterns/non-functional-requirements-investigation-guide.md:103]: I believe @zmmille2 was suggesting adding a link here to the observability section of the playbook

File: docs/design/design-patterns/non-functional-requirements-investigation-guide.md [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+In the preceding example, 'Susan' is responsible for `Parent1` and all of its descendants _except_ for `Parent1/child2` and `Parent1/child2/grandchild1`  (for which 'John' is the stakeholder).  'Sam' is responsible for the entirety of `Parent2` and all of its descendants.
+
+This approach permits the retention of the logical hierarchy of elements themselves while also flexibly interleaving the 'stakeholder' identifications within the hierarchy of topics if/when they may need to diverge due to e.g., CUSTOMER organizational nuances.
+
+***
+
+## Areas of Investigation
+
+* Enterprise Security
```
Comment (fnocera) [inline docs/design/design-patterns/non-functional-requirements-investigation-guide.md:49]: I believe @zmmille2 was suggesting adding a link here to the security section of the playbook (although less complete than some areas)

File: docs/design/design-patterns/non-functional-requirements-investigation-guide.md [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+  * Platform/OS
+    * Hardened
+    * Approved base images
+    * Image repository
+  * Deployment longevity/volatility
+    * Automation
+    * Reproducibility
+      * IaC
+      * Scripting
+      * Other???
```
Comment (fnocera) [inline docs/design/design-patterns/non-functional-requirements-investigation-guide.md:146]: Nit: remove the `???` here and in other parts, I think Other is enough to encourage folks to add/ask abotu additional areas

File: docs/observability/README.md [hunk 10]
```diff
@@ -24,6 +24,7 @@ Building observable systems enables development teams at CSE to measure how well
 - [Tooling and Patterns](tools/README.md)
 - [Observability As Code](observability-as-code.md)
 - [Recommended Practices](best-practices.md)
+- [Non-Functional Requiurements Guidance](../design/design-patterns/non-functional-requirements-investigation-guide.md)
```
Comment (fnocera) [inline docs/observability/README.md:27]: Typo "Requirements" and suggest this is added to a new section at the bottom of the page called "Useful Links" vs in the main structure of the Observability section as this reflects the structure of observability itself.

File: docs/security/README.md [hunk 11]
```diff
@@ -13,6 +13,7 @@ When requesting a security review for your application, please make sure you hav
 * [Secure Coding Practices Quick Reference](https://owasp.org/www-pdf-archive/OWASP_SCP_Quick_Reference_Guide_v2.pdf)
 * [Web Application Security Quick Reference](https://owasp.org/www-pdf-archive//OWASP_Web_Application_Security_Quick_Reference_Guide_0.3.pdf)
 * [Security Mindset/Creating a Security Program Quick Start](https://github.com/OWASP/Quick-Start-Guide/blob/master/OWASP%20Quick%20Start%20Guide.pdf?raw=true)
+* [Non-Functional Requiurements Guidance](../design/design-patterns/non-functional-requirements-investigation-guide.md)
```
Comment (fnocera) [inline docs/security/README.md:16]: Same comment as above for observability.

File: docs/design/design-patterns/non-functional-requirements-investigation-guide.md [hunk 12]
```diff
@@ -0,0 +1,148 @@
+# Non-Functional Requirements Investigation
```
Comment (fnocera) [inline docs/design/design-patterns/non-functional-requirements-investigation-guide.md:1]: Nit: Suggest "Non-Functional Requirements Capture" guide as we should consider positioning this as a guide to help people capture non-functional requirements in a more consistent way.

File: docs/design/design-patterns/non-functional-requirements-investigation-guide.md [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+
+## Background
```
Comment (fnocera) [inline docs/design/design-patterns/non-functional-requirements-investigation-guide.md:3]: Consider "Goals" vs "Background"

**Review by fnocera — CHANGES_REQUESTED** [review]
Thanks for the contribution @sbohlen!! This is much needed as non-functional requirements are such an essential part of understanding architecture and design requirements + limitations.

File: docs/design/design-patterns/non-functional-requirements-investigation-guide.md [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+  * child1
+  * child2
+
+In the preceding example, 'Susan' is responsible for `Parent1` and all of its descendants _except_ for `Parent1/child2` and `Parent1/child2/grandchild1`  (for which 'John' is the stakeholder).  'Sam' is responsible for the entirety of `Parent2` and all of its descendants.
+
+This approach permits the retention of the logical hierarchy of elements themselves while also flexibly interleaving the 'stakeholder' identifications within the hierarchy of topics if/when they may need to diverge due to e.g., CUSTOMER organizational nuances.
+
+***
+
+## Areas of Investigation
```
Comment (fnocera) [inline docs/design/design-patterns/non-functional-requirements-investigation-guide.md:47]: I think this could even be pulled further up, above Investigation and Structure of Outline as it's the most valuable part of the content.

File: docs/design/README.md (hunk 4 above)
Comment (sbohlen) [reply docs/design/README.md:18]: Done.

File: docs/design/design-patterns/non-functional-requirements-investigation-guide.md (hunk 7 above)
Comment (sbohlen) [reply docs/design/design-patterns/non-functional-requirements-investigation-guide.md:103]: Done.

File: docs/design/design-patterns/non-functional-requirements-investigation-guide.md (hunk 8 above)
Comment (sbohlen) [reply docs/design/design-patterns/non-functional-requirements-investigation-guide.md:49]: Done.

File: docs/design/design-patterns/non-functional-requirements-investigation-guide.md (hunk 9 above)
Comment (sbohlen) [reply docs/design/design-patterns/non-functional-requirements-investigation-guide.md:146]: Done.

File: docs/observability/README.md (hunk 10 above)
Comment (sbohlen) [reply docs/observability/README.md:27]: Done,

File: docs/design/design-patterns/non-functional-requirements-investigation-guide.md (hunk 12 above)
Comment (sbohlen) [reply docs/design/design-patterns/non-functional-requirements-investigation-guide.md:1]: Done,

File: docs/design/design-patterns/non-functional-requirements-investigation-guide.md (hunk 13 above)
Comment (sbohlen) [reply docs/design/design-patterns/non-functional-requirements-investigation-guide.md:3]: Done.

File: docs/design/design-patterns/non-functional-requirements-investigation-guide.md (hunk 14 above)
Comment (sbohlen) [reply docs/design/design-patterns/non-functional-requirements-investigation-guide.md:47]: Good idea; done.

File: docs/security/README.md (hunk 11 above)
Comment (sbohlen) [reply docs/security/README.md:16]: Done.

File: docs/design/design-patterns/non-functional-requirements-investigation-guide.md (hunk 6 above)
Comment (sbohlen) [reply docs/design/design-patterns/non-functional-requirements-investigation-guide.md:5]: Good points; done.

File: docs/design/design-patterns/non-functional-requirements_investigation_guide.md (hunk 1 above)
Comment (sbohlen) [reply docs/design/design-patterns/non-functional-requirements_investigation_guide.md]: Added newlines for each sentence per request.  With reduced verbiage, no longer a significant edit task :)

File: docs/design/design-patterns/non-functional-requirements_investigation_guide.md (hunk 3 above)
Comment (sbohlen) [reply docs/design/design-patterns/non-functional-requirements_investigation_guide.md]: Closing/resolving as-is (new edits).   If other inter-document linkages were intended by comment, please open add'l/new comment thread accordingly  --  or of course just re-open this one :)

File: docs/design/design-patterns/non-functional-requirements-investigation-guide.md (hunk 5 above)
Comment (sbohlen) [reply docs/design/design-patterns/non-functional-requirements-investigation-guide.md]: Done.

File: docs/design/design-reviews/recipes/engagement-process.md [hunk 15]
```diff
@@ -12,7 +12,7 @@ The following guidelines outline how Microsoft and the customer together can inc
 
 ## Envisioning / Architecture Design Session (ADS)
 
-Early in an engagement Microsoft works with customers to understand their unique goals and objectives and establish a definition of done. Microsoft dives deep into existing customer infrastructure and architecture to understand potential constraints.
+Early in an engagement Microsoft works with customers to understand their unique goals and objectives and establish a definition of done. Microsoft dives deep into existing customer infrastructure and architecture to understand potential constraints. Recommend leveraging [Non-Functional Requirements Guidance](../../design-patterns/non-functional-requirements-capture-guide.md) to assist.
```
Comment (fnocera) [inline docs/design/design-reviews/recipes/engagement-process.md:15]: Nit: 
```suggestion
Early in an engagement Microsoft works with customers to understand their unique goals and objectives and establish a definition of done. Microsoft dives deep into existing customer infrastructure and architecture to understand potential constraints. Additionally, we seek to understand and uncover specific [non-functional requirements](../../design-patterns/non-functional-requirements-capture-guide.md) that the customer might have.
```

File: docs/design/design-patterns/non-functional-requirements-capture-guide.md [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+In software engineering projects, important characteristics of the system providing for necessary e.g., testability, reliability, scalability, observability, securability, manageability are best considered as first-class citizens in the requirements gathering process.
+By defining these non-functional requirements in detail early in the engagement, they can be properly evaluated when the cost of their impact on subsequent design decisions is comparatively low.
+
+To support the process of capturing a project's _comprehensive_ non-functional requirements, this document offers a taxonomy for non-functional requirements and provides a framework for their identification, exploration, assignment of customer stakeholders, and eventual codification into formal engineering requirements as input to subsequent solution design.
+
+---
+
+## Areas of Investigation
+
+- [Enterprise Security](../../security/README.md)
```
Comment (fnocera) [inline docs/design/design-patterns/non-functional-requirements-capture-guide.md:14]: Nit: I can see a gap between Enterprise Security and the next indented section of content, is there any way to remove this? Suggest looking at how we can make this as clear as possible formatting wise.

**Review by fnocera — APPROVED** [review]
Thanks for integrating the changes @sbohlen, looking great and I am sure this will be super helpful to many teams. 
Left a couple more small nits!

File: docs/design/design-reviews/recipes/engagement-process.md (hunk 15 above)
Comment (sbohlen) [reply docs/design/design-reviews/recipes/engagement-process.md:15]: Re-worked; a mix of your approach and further editorial license from my end :)

File: docs/design/design-patterns/non-functional-requirements-capture-guide.md (hunk 16 above)
Comment (sbohlen) [reply docs/design/design-patterns/non-functional-requirements-capture-guide.md:14]: If I'm understanding the comment correctly, you are calling attn to a repeating rendering oddity that seems to manifest in _all_ elements at that same top-most outline level e.g., `Enterprise Security`, `Enterprise AuthN/AuthZ`, `Enterprise Monitoring/Operations`, `Other standard Enterprise technologies/practices`, `Production Ecosystem`.

If there is a 'trick' to control this in markdown, happy to introduce it, but as far as I can tell, this is an in-built behavior of markdown renderers (?)

Alternately, perhaps a more visually appealing strategy might be to have those top-most elements in the outline actually be separate third-level headings instead of the whole thing being just a single perpetual outline structure.  Something along the lines of...

## Areas of Investigation

### [Enterprise Security](../../security/README.md)
  - Privacy
    - PII
    - HIPAA
  - Encryption
  - Data mobility
    - at rest
    - in motion
    - in process/memory
...
### Enterprise AuthN/AuthZ
  - Users
  - Services
  - Authorities/directories
...
### [Enterprise Monitoring/Operations](../../observability/README.md)
  - Logging
    - Operations
    - Reporting
    - Audit
  - Monitoring
    - Diagnostics/Alerts
...
### Other standard Enterprise technologies/practices
  - Developer ecosystem
    - Platform/OS
      - Hardened
      - Approved base images
...
### Production ecosystem
  - Platform/OS
    - Hardened
    - Approved base images
...

In the above, you will note that the markdown renderer still places increased whitespace after each "###" heading, but its more visually palettable because they are explicitly _headings_ instead of just "level 1 bullets".

Also, separating what is now one singular over-long outline into what is effectively multiple smaller 'topic/SME-area-focused' outlines might make the content easier for a reader to consume (?)

Thoughts --?

Comment (fnocera) [reply docs/design/design-patterns/non-functional-requirements-capture-guide.md:14]: @sbohlen I really like this suggestion, pulling the top-most elements as headers definitely improves the formatting 💯

Comment (sbohlen) [reply docs/design/design-patterns/non-functional-requirements-capture-guide.md]: Done.

**Review by zmmille2 — APPROVED** [review]
Looks great, thank you for your contribution.
